x/mongo/driver/auth/creds/aws_credential_providers/ec2_provider.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package awscredproviders
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"go.mongodb.org/mongo-driver/x/mongo/driver/auth/internal/aws/credentials"
+)
+
+const (
+	// Ec2ProviderName provides a name of EC2 provider
+	Ec2ProviderName = "Ec2Provider"
+
+	awsEC2URI       = "http://169.254.169.254/"
+	awsEC2RolePath  = "latest/meta-data/iam/security-credentials/"
+	awsEC2TokenPath = "latest/api/token"
+)
+
+// An Ec2Provider retrieves credentials from EC2 metadata.
+type Ec2Provider struct {
+	httpClient   *http.Client
+	expiration   time.Time
+	expiryWindow time.Duration
+}
+
+// NewEc2Provider returns a pointer to an EC2 credential provider.
+func NewEc2Provider(httpClient *http.Client, expiryWindow time.Duration) *Ec2Provider {
+	return &Ec2Provider{
+		httpClient:   httpClient,
+		expiration:   time.Time{},
+		expiryWindow: expiryWindow,
+	}
+}
+
+func (e *Ec2Provider) executeAWSHTTPRequest(ctx context.Context, req *http.Request) (io.ReadCloser, error) {
+	const defaultHTTPTimeout = 10 * time.Second
+
+	ctx, cancel := context.WithTimeout(ctx, defaultHTTPTimeout)
+	defer cancel()
+	resp, err := e.httpClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("response failure: %s", resp.Status)
```

Comment:
If the status code is ever not `http.StatusOK`, this may create a connection leak because `resp.Body` will never be closed. Refactor this logic to make sure `resp.Body` is always closed.